security/rbac/engine/engine.go

```diff
@@ -0,0 +1,192 @@
+	// determine whether or not an RPC call is authorized.
+	DecisionUnknown
+)
+
+// String returns the string representation of a Decision object.
+func (d Decision) String() string {
+	return [...]string{"DecisionAllow", "DecisionDeny", "DecisionUnknown"}[d]
+}
+
+// AuthorizationDecision is the output of CEL engine.
+// If decision is allow or deny, policyNames will either contain the names of
+// all the policies matched in the engine that permitted the action, or be
+// empty as the decision was made after all conditions evaluated to false.
+// If decision is unknown, policyNames will contain the list of policies that
+// evaluated to unknown.
+type AuthorizationDecision struct {
+	decision    Decision
+	policyNames []string
+}
+
+// Converts an expression to a parsed expression, with SourceInfo nil.
+func exprToParsedExpr(condition *expr.Expr) *expr.ParsedExpr {
+	return &expr.ParsedExpr{Expr: condition}
+}
+
+// Converts an expression to a CEL program.
+func exprToProgram(condition *expr.Expr) *cel.Program {
+	// TODO(@ezou): implement the conversion from expr to CEL program.
+	return nil
+}
+
+// Returns whether or not a policy is matched.
+func matchesCondition(condition *cel.Program, evalMap map[string]interface{}) (bool, error) {
+	// TODO(@ezou): implement the matching logic using CEL library.
+	return false, nil
+}
+
+// rbacEngine is the struct for an engine created from one RBAC proto.
+type rbacEngine struct {
+	action     pb.RBAC_Action
+	conditions map[string]*cel.Program
+}
+
+// Creates a new rbacEngine from an RBAC.
+func newRbacEngine(rbac pb.RBAC) rbacEngine {
+	action := rbac.Action
+	conditions := make(map[string]*cel.Program)
+	for policyName, policy := range rbac.Policies {
+		conditions[policyName] = exprToProgram(policy.Condition)
+	}
+	return rbacEngine{action, conditions}
+}
+
+// Returns whether a set of authorization arguments, as contained in evalMap, match an
+//  rbacEngine's conditions.
+// If any policy matches, the engine has been matched, and the list of matching policy
+//  names will be returned.
+// Else if any policy is missing attributes, the engine is not matched, and the list of
+//  policy names that can't be evaluated due to missing attributes will be returned.
+// Else, there is not a match.
+func (engine rbacEngine) matches(evalMap map[string]interface{}) (bool, []string) {
+	matchingPolicyNames := []string{}
+	unknownPolicyNames := []string{}
+	for policyName, condition := range engine.conditions {
+		match, err := matchesCondition(condition, evalMap)
+		if err != nil {
+			unknownPolicyNames = append(unknownPolicyNames, policyName)
+		} else if match {
+			matchingPolicyNames = append(matchingPolicyNames, policyName)
+		}
+	}
+	if len(matchingPolicyNames) > 0 {
+		return true, matchingPolicyNames
+	}
+	return false, unknownPolicyNames
+}
+
+// CelEvaluationEngine is the struct for CEL engine.
+type CelEvaluationEngine struct {
```

Comment:
s/CelEvaluationEngine/AuthorizationEngine

Basically one AuthorizationEngine maps to 1 or 2 celEngine